lib/block_view_stake.go

```diff
@@ -0,0 +1,2600 @@
+	"sort"
+)
+
+//
+// TYPES: StakeEntry
+//
+
+type StakeEntry struct {
+	StakeID          *BlockHash
+	StakerPKID       *PKID
+	ValidatorPKID    *PKID
+	StakeAmountNanos *uint256.Int
+	ExtraData        map[string][]byte
+	isDeleted        bool
+}
+
+type StakeMapKey struct {
+	ValidatorPKID PKID
+	StakerPKID    PKID
+}
+
+func (stakeEntry *StakeEntry) Copy() *StakeEntry {
+	return &StakeEntry{
+		StakeID:          stakeEntry.StakeID.NewBlockHash(),
+		StakerPKID:       stakeEntry.StakerPKID.NewPKID(),
+		ValidatorPKID:    stakeEntry.ValidatorPKID.NewPKID(),
+		StakeAmountNanos: stakeEntry.StakeAmountNanos.Clone(),
+		ExtraData:        copyExtraData(stakeEntry.ExtraData),
+		isDeleted:        stakeEntry.isDeleted,
+	}
+}
+
+func (stakeEntry *StakeEntry) Eq(other *StakeEntry) bool {
+	return stakeEntry.StakeID.IsEqual(other.StakeID)
+}
+
+func (stakeEntry *StakeEntry) ToMapKey() StakeMapKey {
+	return StakeMapKey{
+		StakerPKID:    *stakeEntry.StakerPKID,
+		ValidatorPKID: *stakeEntry.ValidatorPKID,
+	}
+}
+
+func (stakeEntry *StakeEntry) RawEncodeWithoutMetadata(blockHeight uint64, skipMetadata ...bool) []byte {
+	var data []byte
+	data = append(data, EncodeToBytes(blockHeight, stakeEntry.StakeID, skipMetadata...)...)
+	data = append(data, EncodeToBytes(blockHeight, stakeEntry.StakerPKID, skipMetadata...)...)
+	data = append(data, EncodeToBytes(blockHeight, stakeEntry.ValidatorPKID, skipMetadata...)...)
+	data = append(data, EncodeUint256(stakeEntry.StakeAmountNanos)...)
+	data = append(data, EncodeExtraData(stakeEntry.ExtraData)...)
+	return data
+}
+
+func (stakeEntry *StakeEntry) RawDecodeWithoutMetadata(blockHeight uint64, rr *bytes.Reader) error {
+	var err error
+
+	// StakeID
+	stakeEntry.StakeID, err = DecodeDeSoEncoder(&BlockHash{}, rr)
+	if err != nil {
+		return errors.Wrapf(err, "StakeEntry.Decode: Problem reading StakeID: ")
+	}
+
+	// StakerPKID
+	stakeEntry.StakerPKID, err = DecodeDeSoEncoder(&PKID{}, rr)
+	if err != nil {
+		return errors.Wrapf(err, "StakeEntry.Decode: Problem reading StakerPKID: ")
+	}
+
+	// ValidatorPKID
+	stakeEntry.ValidatorPKID, err = DecodeDeSoEncoder(&PKID{}, rr)
+	if err != nil {
+		return errors.Wrapf(err, "StakeEntry.Decode: Problem reading ValidatorPKID: ")
+	}
+
+	// StakeAmountNanos
+	stakeEntry.StakeAmountNanos, err = DecodeUint256(rr)
+	if err != nil {
+		return errors.Wrapf(err, "StakeEntry.Decode: Problem reading StakeAmountNanos: ")
+	}
+
+	// ExtraData
+	stakeEntry.ExtraData, err = DecodeExtraData(rr)
+	if err != nil {
+		return errors.Wrapf(err, "StakeEntry.Decode: Problem reading ExtraData: ")
+	}
+
+	return err
+}
+
+func (stakeEntry *StakeEntry) GetVersionByte(blockHeight uint64) byte {
+	return 0
+}
+
+func (stakeEntry *StakeEntry) GetEncoderType() EncoderType {
+	return EncoderTypeStakeEntry
+}
+
+//
+// TYPES: LockedStakeEntry
+//
+
+type LockedStakeEntry struct {
+	LockedStakeID       *BlockHash
+	StakerPKID          *PKID
+	ValidatorPKID       *PKID
+	LockedAmountNanos   *uint256.Int
```

Comment:
I like that we're starting to use uint256'es everywhere. Will be good to eventually convert everything to this...